packages/agent/src/certificate.ts

```diff
@@ -172,7 +208,50 @@ export class Certificate {
   }
 
   public lookup(path: Array<ArrayBuffer | string>): ArrayBuffer | undefined {
-    return lookup_path(path, this.cert.tree);
+    // constrain the type of the result, so that empty HashTree is undefined
+    return lookupResultToBuffer(lookup_path(path, this.cert.tree));
+  }
+
+  public lookup_label(label: ArrayBuffer): ArrayBuffer | HashTree | undefined {
+    return this.lookup([label]);
+  }
+
+  public cache_node_keys(root_key?: Uint8Array): SubnetStatus {
+    const tree = this.cert.tree;
+    let delegation = this.cert.delegation;
+    // On local replica, with System type subnet, there is no delegation
+    if (!delegation && typeof root_key !== 'undefined') {
+      delegation = {
+        subnet_id: Principal.selfAuthenticating(root_key).toUint8Array(),
+        certificate: new ArrayBuffer(0),
+      };
+    }
+    // otherwise use default NNS subnet id
+    else if (!delegation) {
+      delegation = {
+        subnet_id: Principal.fromText(
+          'tdb26-jop6k-aogll-7ltgs-eruif-6kk7m-qpktf-gdiqx-mxtrf-vb5e6-eqe',
+        ).toUint8Array(),
+        certificate: new ArrayBuffer(0),
+      };
+    }
+    const nodeTree = lookup_path(['subnet', delegation?.subnet_id as ArrayBuffer, 'node'], tree);
+    const nodeForks = flatten_forks(nodeTree as HashTree) as HashTree[];
+    nodeForks.length;
+
+    this.#nodeKeys = nodeForks.map(fork => {
```

Comment:
Here `nodeKeys` is an array of public keys. 
When we verify the node signature, we know the node id and would like to find the corresponding public. For this purpose, should `nodeKeys`  be a map from node id to public key?